rustc_interface: Add a new query `pre_configure`

petrochenkov opened the pull request:
It partially expands crate attributes before the main expansion pass (without modifying the crate), and the produced preliminary crate attribute list is used for querying a few attributes that are required very early.

Crate-level cfg attributes on the crate itself are then expanded normally during the main expansion pass, like attributes on any other nodes.
This is a continuation of https://github.com/rust-lang/rust/pull/92473 and one more step to very unstable crate-level proc macro attributes maybe actually working.

Previously crate attributes were pre-configured simultaneously with feature extraction, and then written directly into `ast::Crate`.

petrochenkov committed:
expand: Pass `ast::Crate` by reference to AST transforming passes

Also some more attributes are passed by reference.

petrochenkov commented:
@bjorn3 
I remember your PR introducing the `DEPRECATED_CFG_ATTR_CRATE_TYPE_NAME` lint, I didn't think about it in too much detail back then, but now I'm not sure that link makes much sense.

There are 10 attributes that are used as early as `crate_type` and `crate_name`
```
compiler_builtins
crate_name
crate_type
feature
no_core
no_implicit_prelude
no_std
plugin
recursion_limit
register_tool
```
, i.e. before the crate root is actually expanded by the main expansion pass (`expand_crate`).
(See uses of `pre_configured_attrs` in this PR.)

I don't think it's feasible to delay use of all (*) of these attributes to post-expansion.
So we won't be able to get rid of this "pre-configuration" as a concept.
So we can continue using it for `crate_type` and `crate_name` as well, and do not produce warnings.

However, we need to produce an error if any of these attributes appears after expansion of a crate-level procedural macro attribute.

(*) *Some* of them could potentially be delayed though. It would be nice, because in that case they would work even if produced by crate-level procedural macros.

bjorn3 commented:
crate_name and crate_type affect what the output files are and as such need to be known by the driver. This for example results in us having to modify the Session after it was constructed. If we push back incr comp much further this is going to hurt I think. Disallowing incr comp allows us to scan for crate_type amd crate_name attributes before creating the Session. All other attributes you listed only affect the content of the compilation and not what files are produced or the Session.

Another reason is gccrs which can't really support conditional crate_name and crate_type because as I understand it to the driver needs to know these things before invoking the actual compiler executable to pass the right arguments.

petrochenkov commented:
This should probably be run through crater.
@bors try

bors commented:
:hourglass: Trying commit 13a91c09c660eba2434e10b49b8cc99795d085cb with merge 8df314f433990f5163039edae8245e89581f72ca...
<!-- homu: {"type":"TryBuildStarted","head_sha":"13a91c09c660eba2434e10b49b8cc99795d085cb","merge_sha":"8df314f433990f5163039edae8245e89581f72ca"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4213293368/jobs/7313921031)
Build commit: 8df314f433990f5163039edae8245e89581f72ca (`8df314f433990f5163039edae8245e89581f72ca`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4213293368/jobs/7313921031"},"merge_sha":"8df314f433990f5163039edae8245e89581f72ca"} -->

petrochenkov commented:
@craterbot check

michaelwoerister commented:
Thanks for the PR, @petrochenkov! I'll take a closer look shortly.

petrochenkov commented:
@rust-lang/infra 
*A lot* of "No space left on device" errors in this crater report.
Not sure whether it's expected or not.

petrochenkov commented:
Quite a few regressions from https://github.com/rust-lang/rust/pull/108221/commits/4cf92c4bf66771b9aea796dd71d6646c3a5f9677 (i.e. the fix for https://github.com/rust-lang/rust/issues/104633).

It's mostly the "unknown feature" errors on fully unconfigured crates.

The second source of regressions is that fully unconfigured crates with `#![no_std]` annotations are considered `no_std` now and produce errors like "`#[panic_handler]` function required, but not found" because the panic handler is obviously configured away together with everything else.

So, in practice the attributes from the list in https://github.com/rust-lang/rust/pull/108221#issuecomment-1435760517 aren't supposed to be accounted for in fully-unconfigured crates, I guess.

petrochenkov commented:
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/binary-tree-0.2.0/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/daemon-engine-0.6.0/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/ecs-0.23.1/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/find-winsdk-0.2.0/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/hotg-runicos-base-wasm-0.11.3/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/lazy_static-1.1.1/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/merkle_light-0.4.0/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/promising-future-0.2.4/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/ruspiro-boot-0.5.4/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/ruspiro-cache-0.4.1/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/rwcell-0.1.2/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/tear-0.5.1/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/tokio-async-await-0.1.7/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/tuifw-screen-winapi-0.21.1/log.txt
https://crater-reports.s3.amazonaws.com/pr-108221/try%238df314f433990f5163039edae8245e89581f72ca/reg/vswhere-0.1.0/log.txt

plus almost everything in the `build failed (unknown)` group.

petrochenkov commented:
https://github.com/rust-lang/rust/issues/104633#issuecomment-1437198729

>I'll try to implement this left-to-right amendment for a second crater run

petrochenkov commented:
@bors try

bors commented:
:hourglass: Trying commit fc0c1168c0e9f9236ae4beb0b71ae50e1ae26856 with merge 7e1e4bd31dffeb8c55b00904f3a336f649dd2b84...
<!-- homu: {"type":"TryBuildStarted","head_sha":"fc0c1168c0e9f9236ae4beb0b71ae50e1ae26856","merge_sha":"7e1e4bd31dffeb8c55b00904f3a336f649dd2b84"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4226821794/jobs/7342348225)
Build commit: 7e1e4bd31dffeb8c55b00904f3a336f649dd2b84 (`7e1e4bd31dffeb8c55b00904f3a336f649dd2b84`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4226821794/jobs/7342348225"},"merge_sha":"7e1e4bd31dffeb8c55b00904f3a336f649dd2b84"} -->

petrochenkov commented:
@craterbot check p=1 crates=https://crater-reports.s3.amazonaws.com/pr-108221/retry-regressed-list.txt

michaelwoerister commented:
> If we push back incr comp much further this is going to hurt I think.

As a side note: I don't think there is an intrinsic requirement that incr. comp. needs to use the crate name for identifying cache directories anymore. The initial versions of incr. comp. would read data from the cache directories of upstream crates (for which purpose the crate name was useful) but I'm pretty sure that part has been removed before incr. comp. was stabilized. So with some refactoring effort we probably could make incr. comp. be independent of the crate name.

bjorn3 commented:
We need the crate name and crate type to produce the StableCrateId of the local crate. Serializing the incremental cache uses the StableCrateId for serializing CrateNum. And the CStore needs to be initialized with the StableCrateId. Currently `find_crate_name` is used for that, but that doesn't respect macro expansions already. We also use the crate type at various places during compilation to determine what to do.

> I don't think there is an intrinsic requirement that incr. comp. needs to use the crate name for identifying cache directories anymore.

This is necessary for sharing the incr comp dir between crates, right? Cargo does this. It passes `-C incremental=/path/to/project/target/debug/incremental` and expects rustc to figure out where to place it's incr comp cache in this directory.

bjorn3 commented:
> Currently find_crate_name is used for that, but that doesn't respect macro expansions already.

Turns out `#![crate_name]` in macro expansions is entirely ignored except by the future compat lint. Not even the error when `--crate-name` and `#![crate_name]` disagree is triggered.

```
$ echo '#![cfg_attr(unix, crate_name = "foo")] #![crate_type = "lib"]' | rustc +nightly - -Adeprecated_cfg_attr_crate_type_name
$ ls
librust_out.rlib
$ rustc +nightly -Zls librust_out.rlib
Crate info:
name rust_out
[...]
```

petrochenkov commented:
>Turns out `#![crate_name]` in macro expansions is entirely ignored except by the future compat lint. 

This PR fixes it though.

Also "macro expansion" -> "cfg expansion".
If true crate-level attribute macro expands to any attribute from the list in https://github.com/rust-lang/rust/pull/108221#issuecomment-1435760517 it will behave incorrectly.
We need to prohibit this case, but it's not urgent because crate-level macro attributes are very unstable.

michaelwoerister commented:
> We need the crate name and crate type to produce the StableCrateId of the local crate. Serializing the incremental cache uses the StableCrateId for serializing CrateNum. [...]

Ah yes, it's quite possible that StableCrateId's introduce an intrinsic requirement for the crate name to be known very early. I didn't look into this in any detail.

> > I don't think there is an intrinsic requirement that incr. comp. needs to use the crate name for identifying cache directories anymore.
> 
> This is necessary for sharing the incr comp dir between crates, right? Cargo does this. It passes `-C incremental=/path/to/project/target/debug/incremental` and expects rustc to figure out where to place it's incr comp cache in this directory.

This is what I had in mind: `rustc` only needs some kind of predictable name for each crate's directory inside of `/path/to/project/target/debug/incremental`. That could probably be computed just from command-line arguments and does not have to depend on the actual crate name.

bjorn3 commented:
> That could probably be computed just from command-line arguments and does not have to depend on the actual crate name.

What commandline arguments are suitable for that? The source path is not stable enough and if `--crate-name` and `-Cmetadata` aren't used, `#![crate_name]` is the only way to disambiguate that I know of.

michaelwoerister commented:
> What commandline arguments are suitable for that? The source path is not stable enough and if `--crate-name` and `-Cmetadata` aren't used, `#![crate_name]` is the only way to disambiguate that I know of.

A combination of absolute path of the root source file, `--crate-name`, `-Cmetadata`, `--test`, and maybe `--cfg` is what I was thinking of. But only for coming up a unique directory name within the incr. comp. cache directory, not as a replacement for `StableCrateId` or something like.

If `rustc` is being used to compile the same source twice without `-Cmetadata` then that would indeed not work -- but I'd expect that to already cause symbol conflicts and similar problems.

bjorn3 commented:
> A combination of absolute path of the root source file

That would mean the incr comp cache would be invalidated when you move a project. I think that will only have an effect when cargo already determined that the crate needs to be rebuilt, so maybe that is acceptable?

> If rustc is being used to compile the same source twice without -Cmetadata then that would indeed not work -- but I'd expect that to already cause symbol conflicts and similar problems.

It works if both crates are binaries or if they are never linked together. Or if one is a binary and the other a library (even if the binary links the library as the `StableCrateId` has an is_exe flag as input)

petrochenkov commented:
Some weird "`#[panic_handler]` function required, but not found" errors, I need to investigate.
@rustbot author

petrochenkov commented:
I cannot reproduce the `#[panic_handler]` failures locally.
Perhaps there's some mismatch in environment setup between me and crater, but I don't know what it is exactly.

What is weird is that the affected crates don't even have crate-level cfg attributes.

petrochenkov commented:
@bors try

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4355038879/jobs/7614320293)
Build commit: 236fdbf8f5948d0f2412027726b390e70da6715f (`236fdbf8f5948d0f2412027726b390e70da6715f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4355038879/jobs/7614320293"},"merge_sha":"236fdbf8f5948d0f2412027726b390e70da6715f"} -->

petrochenkov commented:
@craterbot check p=1 crates=https://crater-reports.s3.amazonaws.com/pr-108221/retry-regressed-list.txt

petrochenkov commented:
@bors try

bors commented:
:hourglass: Trying commit 061bb3e7102302aba95213b246881e9f9807c637 with merge 20325169bdf5e1b29e02b22226b88e17ab456c85...
<!-- homu: {"type":"TryBuildStarted","head_sha":"061bb3e7102302aba95213b246881e9f9807c637","merge_sha":"20325169bdf5e1b29e02b22226b88e17ab456c85"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4356533546/jobs/7616890070)
Build commit: 20325169bdf5e1b29e02b22226b88e17ab456c85 (`20325169bdf5e1b29e02b22226b88e17ab456c85`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4356533546/jobs/7616890070"},"merge_sha":"20325169bdf5e1b29e02b22226b88e17ab456c85"} -->

petrochenkov commented:
@craterbot check p=1 crates=https://crater-reports.s3.amazonaws.com/pr-108221/retry-regressed-list.txt

petrochenkov commented:
>I cannot reproduce the `#[panic_handler]` failures locally.
>Perhaps there's some mismatch in environment setup between me and crater, but I don't know what it is exactly.

The mismatch is that crater somehow ignores the `.cargo/config.toml` files.
Reproduced the issue locally now.

petrochenkov commented:
@bors try

bors commented:
:hourglass: Trying commit 1740e3e2c4808ec2f94df349bbc710426d463b05 with merge 5f5972046c11ccf9e9cf93f09b984b927b850de9...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1740e3e2c4808ec2f94df349bbc710426d463b05","merge_sha":"5f5972046c11ccf9e9cf93f09b984b927b850de9"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4410107506/jobs/7729352629)
Build commit: 5f5972046c11ccf9e9cf93f09b984b927b850de9 (`5f5972046c11ccf9e9cf93f09b984b927b850de9`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4410107506/jobs/7729352629"},"merge_sha":"5f5972046c11ccf9e9cf93f09b984b927b850de9"} -->

petrochenkov commented:
@craterbot check p=1 crates=https://crater-reports.s3.amazonaws.com/pr-108221/retry-regressed-list.txt

petrochenkov committed:
Add some tests for the current `#![cfg(FALSE)]` crate behavior

petrochenkov committed:
rustc_interface: Add a new query `pre_configure`

It partially expands crate attributes before the main expansion pass (without modifying the crate), and the produced preliminary crate attribute list is used for querying a few attributes that are required very early.

Crate-level cfg attributes are then expanded normally during the main expansion pass, like attributes on any other nodes.

petrochenkov commented:
All language changes (fix for https://github.com/rust-lang/rust/issues/104633 and related issues) are removed from this PR now.
@rustbot ready

michaelwoerister commented:
@bors r+

bors commented:
:pushpin: Commit bf00e7da03723fdce3045da0ba6ba0e0f1b1ca3f has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister bf00e7da03723fdce3045da0ba6ba0e0f1b1ca3f -->
<!-- homu: {"type":"Approved","sha":"bf00e7da03723fdce3045da0ba6ba0e0f1b1ca3f","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Dylan-DPC commented:
likely failure of [rollup](https://github.com/rust-lang/rust/pull/109450#issuecomment-1478343549)

going to try merging this on its own
@bors rollup=iffy

bors commented:
:hourglass: Testing commit bf00e7da03723fdce3045da0ba6ba0e0f1b1ca3f with merge b7a6709531156f23c00eb11c8f4252a353364cb3...
<!-- homu: {"type":"BuildStarted","head_sha":"bf00e7da03723fdce3045da0ba6ba0e0f1b1ca3f","merge_sha":"b7a6709531156f23c00eb11c8f4252a353364cb3"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4487872494/jobs/7892221202)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/4487872494/jobs/7892221202","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`arm-android`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4487872494/jobs/7891786832) [(plain)](https://github.com/rust-lang-ci/rust/commit/b7a6709531156f23c00eb11c8f4252a353364cb3/checks/12185024757/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling core v0.0.0 (/checkout/library/core)
error: unknown lint: `fuzzy_provenance_casts`
  --> library/alloc/benches/lib.rs:10:1
   |
10 | #![deny(fuzzy_provenance_casts)]
   |
   |
   = note: the `fuzzy_provenance_casts` lint is unstable
   = help: add `#![feature(strict_provenance)]` to the crate attributes to enable
   = help: add `#![feature(strict_provenance)]` to the crate attributes to enable
   = note: `-D unknown-lints` implied by `-D warnings`
[RUSTC-TIMING] collectionsbenches test:true 0.030
error: could not compile `alloc` due to previous error
warning: build failed, waiting for other jobs to finish...
[RUSTC-TIMING] proc_macro test:true 1.300
```

</details>

petrochenkov commented:
Passed pre-configured attributes to `pre_expansion_lint` too.
@bors r=michaelwoerister

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4490994998/jobs/7902708882)
Approved by: michaelwoerister
Pushing dadbc672cd31440987083fa20e202910f8a2f2d7 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"michaelwoerister","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4490994998/jobs/7902708882"},"merge_sha":"dadbc672cd31440987083fa20e202910f8a2f2d7"} -->

bors commented:
:eyes: Test was successful, but fast-forwarding failed: 422 Update is not a fast forward

lqd commented:
(post-merge perf results on an unmerged PR 🤔 )

Kobzol commented:
I think that the bot only checks this: https://github.com/rust-lang/rust/pull/108221#issuecomment-1479957950 and it doesn't check further if the merge fails for some reason.

petrochenkov commented:
>Test was successful, but fast-forwarding failed: 422 Update is not a fast forward

What is this magic.
I have never seen this issue before, but apparently it's pretty old - https://github.com/rust-lang/rust/issues/43535.

petrochenkov commented:
@bors r=michaelwoerister

bors commented:
:pushpin: Commit d0adaf6422df90a984b6411a71c78dc23c9ff297 has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister d0adaf6422df90a984b6411a71c78dc23c9ff297 -->
<!-- homu: {"type":"Approved","sha":"d0adaf6422df90a984b6411a71c78dc23c9ff297","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit d0adaf6422df90a984b6411a71c78dc23c9ff297 with merge f74d795d3f2b3b91451ddfbfbea953d539d4f299...
<!-- homu: {"type":"BuildStarted","head_sha":"d0adaf6422df90a984b6411a71c78dc23c9ff297","merge_sha":"f74d795d3f2b3b91451ddfbfbea953d539d4f299"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4496011959/jobs/7910427694)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/4496011959/jobs/7910427694","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`i686-mingw-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4496011959/jobs/7910261409) [(plain)](https://github.com/rust-lang-ci/rust/commit/f74d795d3f2b3b91451ddfbfbea953d539d4f299/checks/12209451399/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[2616/3021] Linking CXX static library lib\libLLVMRISCVCodeGen.a
[2617/3021] Linking CXX static library lib\libLLVMSparcCodeGen.a
[2618/3021] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/DebugMap.cpp.obj
[2619/3021] Linking CXX static library lib\libLLVMWebAssemblyUtils.a
FAILED: lib/libLLVMWebAssemblyUtils.a 
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E rm -f lib\libLLVMWebAssemblyUtils.a && C:\a\rust\rust\mingw32\bin\ar.exe qc lib\libLLVMWebAssemblyUtils.a  lib/Target/WebAssembly/Utils/CMakeFiles/LLVMWebAssemblyUtils.dir/WebAssemblyUtilities.cpp.obj lib/Target/WebAssembly/Utils/CMakeFiles/LLVMWebAssemblyUtils.dir/WebAssemblyTypeUtilities.cpp.obj && C:\a\rust\rust\mingw32\bin\ranlib.exe lib\libLLVMWebAssemblyUtils.a && cd ."
C:\a\rust\rust\mingw32\bin\ranlib.exe: could not create temporary file whilst writing archive: no more archived files
[2620/3021] Linking CXX static library lib\libLLVMSystemZCodeGen.a
[2621/3021] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/MachODebugMapParser.cpp.obj
[2622/3021] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/DwarfLinkerForBinary.cpp.obj
[2623/3021] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/dsymutil.cpp.obj
[2623/3021] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/dsymutil.cpp.obj
[2624/3021] Linking CXX executable bin\llvm-profdata.exe
[2625/3021] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/Reproducer.cpp.obj
[2626/3021] Linking CXX static library lib\libLLVMPasses.a
ninja: build stopped: subcommand failed.
command did not execute successfully, got: exit code: 1


build script failed, must exit now', C:\Users\runneradmin\.cargo\registry\src\index.crates.io-1cd66030c949c28d\cmake-0.1.48\src\lib.rs:975:5
 finished in 293.387 seconds
Build completed unsuccessfully in 0:06:06
Build completed unsuccessfully in 0:06:06
make: *** [Makefile:80: ci-mingw-subset-2] Error 1
```

</details>

petrochenkov commented:
@bors retry

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout cratecfg` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self cratecfg --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging compiler/rustc_resolve/src/macros.rs
CONFLICT (content): Merge conflict in compiler/rustc_resolve/src/macros.rs
Auto-merging compiler/rustc_resolve/src/lib.rs
CONFLICT (content): Merge conflict in compiler/rustc_resolve/src/lib.rs
Auto-merging compiler/rustc_expand/src/base.rs
Auto-merging compiler/rustc_builtin_macros/src/test_harness.rs
Auto-merging compiler/rustc_builtin_macros/src/standard_library_imports.rs
CONFLICT (content): Merge conflict in compiler/rustc_builtin_macros/src/standard_library_imports.rs
Auto-merging compiler/rustc_builtin_macros/src/proc_macro_harness.rs
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

bors commented:
:pushpin: Commit aca1b1e0b3820ae03622c37fb1b32e005d737fbf has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister aca1b1e0b3820ae03622c37fb1b32e005d737fbf -->
<!-- homu: {"type":"Approved","sha":"aca1b1e0b3820ae03622c37fb1b32e005d737fbf","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4502219081/jobs/7927247200)
Approved by: michaelwoerister
Pushing df7fd9995f10627f25ccb325f693a11b3395a73c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"michaelwoerister","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4502219081/jobs/7927247200"},"merge_sha":"df7fd9995f10627f25ccb325f693a11b3395a73c"} -->
